File: Java/leetcode75/src/main/java/org/phoenix/leetcode/challenges/Problem02_GreatestCommonDivisorOfStrings.java
package org.phoenix.leetcode.challenges;

public class Problem02_GreatestCommonDivisorOfStrings {
    public String gcdOfStrings(String str1, String str2) {
        if (!(str1 + str2).equals(str2 + str1)) {
            return "";
        }
        int gcdLength = gcdOfLength(str1.length(), str2.length());
        return str1.substring(0, gcdLength);
    }

    private int gcdOfLength(int m, int n) {
        if (n == 0) return m;
        return gcdOfLength(n, (m % n));
    }
}

File: JavaScript/leetcode75/src/main/Problem02_GreatestCommonDivisorOfStrings.js
/**
 * @param {string} str1
 * @param {string} str2
 * @return {string}
 */
var gcdOfStrings = function (str1, str2) {
  if (str1 + str2 !== str2 + str1) return "";

  const gcdLength = gcdOfLength(str1.length, str2.length);
  return str1.slice(0, gcdLength);
};

var gcdOfLength = function (m, n) {
  if (n === 0) return m;
  return gcdOfLength(n, m % n);
};

module.exports = gcdOfStrings;
